src/arithmetic/boolean_funcs/boolean_funcs.c

```diff
@@ -0,0 +1,280 @@
+		char *error;
+		asprintf(&error, "Type mismatch: expected %s but was %s", SIType_ToString(SI_TYPE(a)),
+				 SIType_ToString(SI_TYPE(b)));
+		return SI_Error(error);
+	}
+
+	switch(SI_TYPE(a)) {
+	case T_STRING:
+		return SI_BoolVal(SIValue_Compare(a, b) <= 0);
+	case T_INT64:
+	case T_DOUBLE:
+	case T_BOOL:
+		return SI_BoolVal(SI_GET_NUMERIC(a) <= SI_GET_NUMERIC(b));
+	default:
+		assert(false);
+	}
+}
+
+SIValue AR_EQ(SIValue *argv, int argc) {
+	SIValue a = argv[0];
+	SIValue b = argv[1];
+
+	// Comparisons with NULL values always return NULL.
+	if(SIValue_IsNull(a) || SIValue_IsNull(b)) return SI_NullVal();
+
+	// Non-comparable types cannot equal each other.
+	if(!SI_VALUES_ARE_COMPARABLE(a, b)) return SI_BoolVal(false);
+
+	switch(SI_TYPE(a)) {
+	case T_STRING:
+		return SI_BoolVal(SIValue_Compare(a, b) == 0);
+	case T_INT64:
+	case T_DOUBLE:
+	case T_BOOL:
+		return SI_BoolVal(SI_GET_NUMERIC(a) == SI_GET_NUMERIC(b));
+	default:
+		assert(false);
+	}
+}
+
+SIValue AR_NE(SIValue *argv, int argc) {
+	SIValue a = argv[0];
+	SIValue b = argv[1];
+
+	// Comparisons with NULL values always return NULL.
+	if(SIValue_IsNull(a) || SIValue_IsNull(b)) return SI_NullVal();
+
+	// Non-comparable types cannot equal each other.
+	if(!SI_VALUES_ARE_COMPARABLE(a, b)) return SI_BoolVal(true);
+
+	switch(SI_TYPE(a)) {
+	case T_STRING:
+		return SI_BoolVal(SIValue_Compare(a, b) != 0);
+	case T_INT64:
+	case T_DOUBLE:
+	case T_BOOL:
+		return SI_BoolVal(SI_GET_NUMERIC(a) != SI_GET_NUMERIC(b));
+	default:
+		assert(false);
+	}
+}
+
+void Register_BooleanFuncs() {
+	SIType *types;
+	AR_FuncDesc *func_desc;
+
+	types = array_new(SIType, 2);
+	types = array_append(types, T_BOOL | T_NULL);
+	types = array_append(types, T_BOOL | T_NULL);
+	func_desc = AR_FuncDescNew("and", AR_AND, 2, types);
+	AR_RegFunc(func_desc);
+
+	types = array_new(SIType, 2);
+	types = array_append(types, T_BOOL | T_NULL);
+	types = array_append(types, T_BOOL | T_NULL);
+	func_desc = AR_FuncDescNew("or", AR_OR, 2, types);
+	AR_RegFunc(func_desc);
+
+	types = array_new(SIType, 2);
+	types = array_append(types, T_BOOL | T_NULL);
+	types = array_append(types, T_BOOL | T_NULL);
+	func_desc = AR_FuncDescNew("xor", AR_XOR, 2, types);
+	AR_RegFunc(func_desc);
+
+	types = array_new(SIType, 1);
+	types = array_append(types, T_BOOL | T_NULL);
+	func_desc = AR_FuncDescNew("not", AR_NOT, 1, types);
+	AR_RegFunc(func_desc);
+
+	types = array_new(SIType, 2);
+	types = array_append(types, (SI_NUMERIC | T_STRING | T_BOOL | T_NULL));
+	types = array_append(types, (SI_NUMERIC | T_STRING | T_BOOL | T_NULL));
+	func_desc = AR_FuncDescNew("gt", AR_GT, 2, types);
+	AR_RegFunc(func_desc);
+
+	types = array_new(SIType, 2);
+	types = array_append(types, (SI_NUMERIC | T_STRING | T_BOOL | T_NULL));
+	types = array_append(types, (SI_NUMERIC | T_STRING | T_BOOL | T_NULL));
+	func_desc = AR_FuncDescNew("ge", AR_GE, 2, types);
+	AR_RegFunc(func_desc);
+
+	types = array_new(SIType, 2);
```

Comment:
I was thinking maybe we should introduce an API to describe AR func expectations,

e.g.
```types = NewRequirement()
ExpectType(types, (SI_NUMERIC | T_STRING | T_BOOL | T_NULL));
ExpectType(types, (SI_NUMERIC | T_STRING | T_BOOL | T_NULL));
```
Need to find better naming for this.
Instead of dealing directly with an array.

Reply:
Sounds good, I'll come up with something!

Reply:
I still think this is a good idea, but haven't had time to implement anything.